archive/import.go

```diff
@@ -2,156 +2,344 @@ package archive
+	d := &tsDirWriter{
+		ark:          iw.ark,
+		importWriter: iw,
+		ctx:          iw.ctx,
+		day:          midnight,
+		tsDir:        iw.ark.DataPath.AppendPath(dataDirname, newTsDir(midnight).name()),
+	}
+	if dirmkr, ok := d.ark.dataSrc.(iosrc.DirMaker); ok {
+		if err := dirmkr.MkdirAll(d.tsDir, 0755); err != nil {
+			return nil, err
+		}
+	}
+	return d, nil
+}
+
+func (w *tsDirWriter) addBufSize(add int64) {
+	old := w.bufSize()
+	w.size += add
+	w.importWriter.bufSize += w.bufSize() - old
+}
+
+// bufSize returns the actually buffer size as a multiple of importLZ4BlockSize.
+// This is done to ensure that whenever spill the generated spill is compressed
+// nicely onto disk.
+func (w *tsDirWriter) bufSize() int64 {
+	return (w.size / int64(importLZ4BlockSize)) * int64(importLZ4BlockSize)
+}
+
+// approxTotalBytes is the sum of the size of compressed records spilt to disk
+// and a crude approximation of the buffer record bytes (simply bufBytes / 2).
+func (d *tsDirWriter) approxTotalBytes() int64 {
+	b := (d.bufSize() / 2)
+	if d.spiller != nil {
+		b += d.spiller.SpillSize()
+	}
+	return b
+}
+
+func (d *tsDirWriter) writeOne(rec *zng.Record) error {
+	d.records = append(d.records, rec)
+	d.addBufSize(int64(len(rec.Raw)))
+	if d.approxTotalBytes() > d.ark.LogSizeThreshold {
+		if err := d.flush(); err != nil {
+			return err
+		}
 	}
 	return nil
 }
 
-func (d *importDriver) writeOne(rec *zng.Record) error {
-	if d.dataFileWriter != nil && !d.firstTs.DayOf().Contains(rec.Ts()) {
-		// Don't allow data files to include records from multiple tsDir time spans.
-		if err := d.close(); err != nil {
+func (d *tsDirWriter) spill() error {
+	if len(d.records) == 0 {
+		return nil
+	}
+	if d.spiller == nil {
+		var err error
+		d.spiller, err = spill.NewMergeSort(importCompareFn(d.ark))
+		if err != nil {
 			return err
 		}
+		d.spiller.SetWriterOpts(zngio.WriterOpts{
+			LZ4BlockSize:     importLZ4BlockSize,
+			StreamRecordsMax: importStreamRecordsMax,
+		})
+	}
+	if err := d.spiller.Spill(d.records); err != nil {
+		return err
 	}
-	if d.dataFileWriter == nil {
-		if err := d.newWriter(rec); err != nil {
+	d.records = d.records[:0]
+	d.addBufSize(d.size * -1)
+	return nil
+}
+
+func (d *tsDirWriter) flush() error {
+	var r zbuf.Reader
+	if d.spiller != nil {
+		if err := d.spill(); err != nil {
 			return err
 		}
+		spiller := d.spiller
+		d.spiller, r = nil, spiller
+		defer spiller.Cleanup()
+	} else {
+		// If len of records is 0 and spiller is nil, the tsDirWriter is empty.
+		// Just return nil.
+		if len(d.records) == 0 {
+			return nil
+		}
+		expr.SortStable(d.records, importCompareFn(d.ark))
+		r = zbuf.Array(d.records).NewReader()
+		defer func() {
+			d.records = d.records[:0]
+			d.addBufSize(d.size * -1)
+		}()
+	}
+	w, err := newChunkWriter(d.ctx, d.ark, d.tsDir)
+	if err != nil {
+		return err
+	}
+	if err := zbuf.CopyWithContext(d.ctx, w, r); err != nil {
+		w.close(d.ctx)
```

Comment:
This `w.close(.)` call should be something like `w.abort()`, where chunkWriter.Abort just closes the dataFileWriter & the indexTempWriter, but doesn't try to create the seek file & push that up.

Reply:
Good point.

Reply:
Is/will the archive read path be resilient if there's a chunk with no accompanying index?